// quil/dist/components/view/bands.component.js
import { Component, Input } from '@angular/core';
import { ViewBandVm } from '../viewmodel/view-band-vm';
import { BIService } from '../../services/bi-chart-viewer';
var BandsComponent = /** @class */ (function () {
    function BandsComponent(cService) {
        this.cService = cService;
        this.band = new ViewBandVm();
        console.log('Get Chart complete');
    }
    BandsComponent.prototype.checkifLbl = function (dim) {
        return dim.type.toString() == "Label";
    };
    BandsComponent.prototype.checkifBar = function (dim) {
        return dim.type.toString() == "Bar";
    };
    BandsComponent.prototype.checkifLine = function (dim) {
        return dim.type.toString() == "Line";
    };
    BandsComponent.prototype.checkifPie = function (dim) {
        return dim.type.toString() == "Pie";
    };
    BandsComponent.prototype.checkifDou = function (dim) {
        return dim.type.toString() == "Doughnut";
    };
    BandsComponent.prototype.checkifTable = function (dim) {
        return dim.type.toString() == "Table";
    };
    BandsComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.pBand.charts.forEach(function (element1) {
            _this.cService.GetChartData(_this.pTitle, element1.title).subscribe((function (k) {
                _this.band.dimensions.push(k);
                if (_this.parent) {
                    _this.parent.chartDatas.push(_this);
                }
            }), function (error) { console.log(error); }, function () { console.log('Get Chart complete'); });
        });
    };
    BandsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bands',
                    template: "\n    <div [ngClass]=\"checkifLbl(dim) ? 'ui-lg-3 ui-md-6 ui-g-12' :'ui-lg-6 ui-md-12 ui-g-12 chart-card'\" *ngFor=\"let dim of band.dimensions\">\n      <label-viewer *ngIf=\"checkifLbl(dim)\" [labelModel]=\"dim\"></label-viewer>\n      <table-viewer *ngIf=\"checkifTable(dim)\" [tableModel]=\"dim\"></table-viewer>\n      <chart-viewer *ngIf=\"checkifBar(dim)\" [barModel]=\"dim\"></chart-viewer>\n      <chart-viewer *ngIf=\"checkifLine(dim)\" [lineModel]=\"dim\"></chart-viewer>\n      <chart-viewer *ngIf=\"checkifPie(dim)\" [pieModel]=\"dim\"></chart-viewer>\n      <chart-viewer *ngIf=\"checkifDou(dim)\" [douModel]=\"dim\"></chart-viewer>\n    </div>\n  ",
                    styles: ["\n    @media print{.no-print,.no-print *{display:none !important}}.label-card{max-height:165px;margin-bottom:auto !important}.ellipsis{text-overflow:ellipsis;overflow:hidden;max-width:87%;white-space:nowrap;display:inline-block}.float-right{float:right}.no-margin{margin:auto}.main-value{font-size:36px;font-weight:bold;text-align:center}.caption{font-weight:bold}.table-view-col{width:20px;text-align:center}.ui-datatable-header h4{margin:auto}.ui-datatable .ui-datatable-footer{padding:4px !important}.pad4{padding:4px}.pad-inherit{padding:inherit}.pad05{padding:0px 5px 5px 5px}.marg-top-10{margin-top:10px}@media screen and (max-width: 641px){.c-input span input{width:100% !important}.c-input span{width:100% !important}}.ui-fieldset{padding:2px !important}h2{font-weight:500 !important}.card{margin-bottom:0px !important}.chart-card{padding:0px 5px}.add-btn{width:20px !important;height:20px !important;padding:11px !important;line-height:0em !important;float:right}.inp-lbl{padding:4px}.inp-div .ui-g-8{padding:.2em !important}.ui-tree{width:314px;min-height:40px}.entity-tree .ui-tree{width:44em;max-height:21%;overflow-y:auto}.pull-right{float:right}.ui-dialog{max-height:99% !important}div .dropdown{min-width:20px !important}.ui-dropdown{min-width:75px !important}.add-btn .fa{font-size:15px}.ui-treenode-label{min-width:74%}.pad-left-0{padding-left:0pc}.ui-accordion-header{margin:3px 0px}.remove-button{font-size:21px;background-color:white;color:#555555;display:block;text-align:center;border-radius:13px;width:28px;cursor:pointer;float:right}.remove-button:hover{background-color:#374b69;color:white}.ui-treetable thead th,body .ui-treetable thead tr th{padding:0px}.dropdown div{width:auto !important}.ddfill div{width:100% !important}body .ddfill .ui-dropdown .ui-dropdown-trigger .fa{margin-left:90% !important}.newConDialog .ui-dialog .ui-dialog-content{overflow:inherit !important}.newConDialog .ui-dialog{top:70px !important}.conddwn div{display:block;margin-bottom:15px}body .ui-treetable thead tr{border:1px solid #bdbdbd}\n  "]
                },] },
    ];
    /** @nocollapse */
    BandsComponent.ctorParameters = function () { return [
        { type: BIService, },
    ]; };
    BandsComponent.propDecorators = {
        'pBand': [{ type: Input, args: ['model',] },],
        'pTitle': [{ type: Input, args: ['caption',] },],
        'parent': [{ type: Input, args: ['parent',] },],
    };
    return BandsComponent;
}());
export { BandsComponent };
